README
|========================================================================|
| voronoi v2.2 (2004-05-20)  GIMP plug-in generating Voronoi diagrams    |
| Copyright (C) 2001-2004 Yeti (David Necas) <yeti@physics.muni.cz>      |
|========================================================================|

Voronoi is a GIMP plug-in generating Voronoi diagram of a semi-random set
of points and offering some interesting methods of coloring it.


Installation
------------

Requirements: Gimp 2.0 (use version 1.0 with Gimp 1.2).

Untar the tar file (well I guess you've got that far already), then in the
directory thus created type

      make install

to install the plug-in to the system-wide plug-in directory, or

      make install-user

to install it to your user plug-in directory.  You can uninstall it later by
`make uninstall' or `make uninstall-user', respectively. When something goes
wrong, try modify the Makefile or, if you are really desperate, the source.


Usage
-----

This plug-in installs itself in the Filters/Render/Pattern menu as `Voronoi'.

It renders Voronoi diagrams.  All Voronoi options are either self-explanatory
or extremely crypitc.  You don't need to know what they mean, just try to
change them and you will see.  There are/will be some images demonstrating
effects of various parameters on project home page (see below).

For the curious, a brief explanations follows:

* What is a Voronoi diagram?

  Imagine a finite set of points in some metric space (or just just normal
  n-dimensional space, if you don't know much about metric spaces).  If you
  now take any point of the space, there definitely exists a point of the set
  which is closest to this point.  This way the set breaks the space to parts,
  (called cells below) each consisting of points nearest to a one point of
  the set.  This division is what is called Voronoi diagram of the set
  (roughly).

----- Graphic Options -----
* Horizontally tileable.
* Vertically tileable.

  These options cause the image to be tileable, horizontally and/or
  vertically.

  The conditions of tileability, grid type, and image aspect ratio are
  in general case inconsistent.  Voronoi thus can fulfil them only
  appoximately -- you may notice deformations of large cells, but you
  should not notice anything on smaller scales.

* Coloring method.
* Gradient.

  For RGB images, you can choose between generating grayscale image and
  and coloring the image with some gradient.  For grayscale images, only
  grayscale output is possible, of course.

  Note applying the gradient later, using `Map to gradient' filter, gives
  slightly worse results than doing the same directly in Voronoi, especially
  in case of gradients with very smooth or very abrupt color transitions.

* Invert.

  Whether the gradient (or gray scale) should be mapped inverted.

----- Cells -----
* Random seed.

  Standard GIMP random seed controls.

* Average cell size.

  If the same number of points was placed in regular square grid, this would
  be side of the square.  For tileable grids this is true only approximately.
  The preview scales very large cells down (see the preview title for zoom
  value), but it is inaccurate also in other ways for them.

* Cell type.

  The points can be generated completely randomly (`Random') or in a kind of
  a grid, you can choose between `Square', `Rhomb', `Triangle (horizontal)',
  `Triangle (vertical)', `Hexagonal (horizontal)' and `Hexagonal (vertical)'.
  Note grid type and cell shape are complementary, trigonal grid implies
  hexagonal cells and vice versa.

----- Deformation -----
Make sense only for non-random grids.

* Randomness.

  Amplitude of random displacements of the points from their perfect grid
  positions.  It is proportional to standard deviation of the shifts.

* Correlation.

  Autocorrelation length of the random shifts relative to cell size.
  Larger values mean there exists a correlation between shitfs of near
  cell centers (the grid is deformed `smoothly'), small values mean random
  shifts of different cells are independent.

* Interstitials.
* Vacancies.

  As you know from physics of solid state, these are two dislocation types:
  extra objects placed where no grid point should be (interstitials), and
  missing object where one would expect one (vacancies).

----- Quantities Shown -----
OK, we have generated a Voronoi diagram and what next?  This is a GIMP
plug-in, so we have to draw something, too.  There are several different
quantities that can be drawn, and you can mix them freely (the weights
are relative):

* Random solid color:
  A random solid color is assigned to each cell.

* Radial distance
  Euclidean distance from the pixel to nearest point of the set (cell center).

* Segment distance:
  Euclidean distance from the nearest border line rescaled so that border
  always gets 1 and the cell center itself 0.

* Border distance:
  Plain Euclidean distance from the nearest border line.

* Second nearest distance:
  Euclidean distance from the second nearest object.

* Scalar product:
  Scalar product of vectors pointing to the pixel from first and second
  nearest objects (somewhat shifted).

----- Scale & Shift -----

* Rescale (log).

  A value to multiply the weighted sum of the above quantities by.  (The
  scale is logarithmic if you wonder why its default value is zero.)

* Shift.

  Value to add to the rescaled weighted sum above.


Some details you don't want to know
-----------------------------------

Many.  Many.  Many.


Authors
-------

Yeti (David Necas) <yeti@physics.muni.cz>.

Report bugs to <yeti@physics.muni.cz> (please include `voronoi' in subject).

Web: http://trific.ath.cx/software/gimp-plugins/voronoi/


Copying
-------

Voronoi can be copied and/or modified under the terms of GNU General
Public License.  Please see COPYING for details.


NEWS

Version 2.2 (May 2004)
- Fixed missing/extra border cells in tileable square grids with low
  randomness.
- Due to typo in formula correlation was reducing randomness a way too much,
  this was fixed and larger ranges of randomnes and correlation were allowed.
- Reset properly resets `Invert' state.
- Dialog buttons were reordered to follow HIG and stock Reset is used instead
  Defaults.


Version 2.1 (May 2004)
- Unpredictable randomness effect depending on cell size for small
  correlations was fixed.
- Different quantitiative effect of Interstitials/Vacancies on different
  grid types was fixed.
- It is possible to map the gradient inverted [PDB change].


Version 2.0 (May 2004)
A substantial rewrite, not even nearly PDB level compatible.
- Full support for horizontal/vertical tileability.
- Visualization types can be mixed with arbitrary weights.
- Preview shows real cell size when feasible and displays zoom value when it
  is not.
- Grid dislocations: interstitials and vacancies.
- New visualization type: scalar product.
- New grid types: rhomb, vertical triangle and hexagon.
- Random solid colors are kept consistent when randomization changes.
- `Coupling' parameter replaced with well defined autocorrelation length.
- Various small improvements and bugfixes.
- NEWS split to separate file, README updated, `About' dialog.
Known issues:
- It's slow and bloated.
- Large cells are deformed even when tileability is not required.


Version 1.5 (Apr 2004)
- Fixed random seed initialization to match the new API.
- Added random_seed to PlasmaVals to remember `Randomize' gui state.
- Using GRand to get random numbers.
- Some clean-up.


Version 1.4 (Jan 2004)
- Changed dialog construction and main event loop to work with gimp-1.3.23.
- Some code cleanup.


Version 1.3 (Oct 2003)
- Works with gimp-1.3.21 and newer.


Version 1.2 (Jan 2003)
- Ported to use the new Gimp random seed widget (great!).


Version 1.1 (Oct 2002)
Ported to Gimp-1.3/Gtk-2.0:
- Renamed and retyped stuff as necessary
- Created a brand new preview widget using GtkImage & GdkPixbuf
- Stockized stock items and reordered buttons to Gtk-2.0 order.
  Can't use GIMP_STOCK_RESET, becuase its label associates `Defaults' while
  its image `Revert'
- Added accels to labels.  FIXME: Ugly Gimp random seed doesn't work
- Removed Esc-exits accel, since Gtk-2.0 does it itself


Version 1.0 (Sep 2001)
- This is the first version.  No news, no changes, no known bugs.




voronoi.h
/* @(#) $Id: voronoi.h,v 1.15 2004/05/08 19:13:22 yeti Exp $ */
#ifndef VORONOI_H
#define VORONOI_H

#include <glib.h>

#include <stdio.h>

/* just some small but definitely nonzero number (it should be several orders
   larger than machine epsilon, at least) */
#define EPS 0.0000001
/* how larger the squarized grid should be (measured in squares) */
#define SQBORDER 2

extern GRand *rng;

/* grid type
   (more precisely, neighbourhood type---square grid is really square, but to
   get triangular neighbourhood the grid itself must be hexagonal and vice
   versa; random grid is really random) */
typedef enum {
  VORONOI_GRID_RANDOM = -1,
  VORONOI_GRID_SQUARE = 0,
  VORONOI_GRID_RHOMB = 1,
  VORONOI_GRID_TRIANGLE_H = 2,
  VORONOI_GRID_TRIANGLE_V = 3,
  VORONOI_GRID_HEXAGON_H = 4,
  VORONOI_GRID_HEXAGON_V = 5,
} VoronoiGridType;

typedef enum {
  VORONOI_VALUE_FLAT = 0,
  VORONOI_VALUE_RADIAL = 1,
  VORONOI_VALUE_SEGMENTED = 2,
  VORONOI_VALUE_BORDER = 3,
  VORONOI_VALUE_SECOND = 4,
  VORONOI_VALUE_SPROD = 5,
  VORONOI_VALUE_LAST
} VoronoiValueType;

typedef struct {
  gdouble x, y;
} VoronoiCoords;

typedef struct {
  VoronoiCoords v; /* line equation: v*r == d (XXX: NOT v*r + d == 0) */
  gdouble d;
} VoronoiLine;

typedef struct VoronoiObject {
  VoronoiCoords pos; /* coordinates */
  VoronoiLine rel; /* precomputed coordinates relative to currently processed
                      object and their norm */
  gdouble angle; /* precomputed angle relative to currently processed object
                    (similar as rel) */
  gdouble random;  /* a random number in [0,1], generated to be always the
                      same for the same grid size */
  GSList *ne; /* neighbour list */
} VoronoiObject;

/* convience macro to make the source readable
   use VOBJ(p->next)->angle = M_PI2 */
#define VOBJ(x) ((VoronoiObject*)(x)->data)

#define DOTPROD_SS(a, b) ((a).x*(b).x + (a).y*(b).y)
#define DOTPROD_SP(a, b) ((a).x*(b)->x + (a).y*(b)->y)
#define DOTPROD_PS(a, b) ((a)->x*(b).x + (a)->y*(b).y)
#define DOTPROD_PP(a, b) ((a)->x*(b)->x + (a)->y*(b)->y)

extern gdouble radial_factor;

GSList**        random_grid              (const VoronoiGridType gridtype,
                                          const gint width,
                                          const gint height,
                                          const gdouble side,
                                          const gdouble sigma,
                                          const gdouble lambda,
                                          const gboolean htil,
                                          const gboolean vtil,
                                          const gdouble interstit,
                                          const gdouble vacancies,
                                          gint *wsq,
                                          gint *hsq);
void            destroy_grid_contents    (GSList **squares,
                                          const gint wsq,
                                          const gint hsq);
void            destroy_neighbourhoods   (GSList **squares,
                                          const gint wsq,
                                          const gint hsq);
void            find_voronoi_neighbours  (GSList **squares,
                                          const gint wsq,
                                          const gint hsq);
void            find_voronoi_neighbours_iter  (GSList **squares,
                                               const gint wsq,
                                               const gint hsq,
                                               const gint iter);
/*
void            remove_objects           (GSList **squares,
                                          const gint wsq,
                                          const gint hsq,
                                          gboolean htil,
                                          gboolean vtil);
                                          */
VoronoiObject*  find_owner               (GSList **squares,
                                          const gint wsq,
                                          const gint hsq,
                                          const VoronoiCoords *const point);
VoronoiObject*  move_along_line          (const VoronoiObject *const owner,
                                          const VoronoiCoords *const start,
                                          const VoronoiCoords *const end,
                                          gint *next_safe);
void            neighbourize             (GSList *ne0,
                                          const VoronoiCoords *const center);
void            compute_segment_angles   (GSList *ne0);

typedef gdouble (*VoronoiValueFunc)(const VoronoiCoords* const,
                                    const VoronoiObject* const);

gdouble         value_flat                (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);
gdouble         value_radial              (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);
gdouble         value_segmented           (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);
gdouble         value_border              (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);
gdouble         value_second              (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);
gdouble         value_sprod               (const VoronoiCoords *const point,
                                           const VoronoiObject *const owner);

static const VoronoiValueFunc value_function[] = {
  &value_flat,      /* VORONOI_VALUE_FLAT */
  &value_radial,    /* VORONOI_VALUE_RADIAL */
  &value_segmented, /* VORONOI_VALUE_SEGMENTED */
  &value_border,    /* VORONOI_VALUE_BORDER */
  &value_second,    /* VORONOI_VALUE_SECOND */
  &value_sprod,     /* VORONOI_VALUE_SPROD */
};

/* coords_minus(const VoronoiCoords*const, const VoronoiCoords*const) {{{ */
static inline VoronoiCoords
coords_minus(const VoronoiCoords *const a, const VoronoiCoords *const b)
{
  VoronoiCoords z;

  z.x = a->x - b->x;
  z.y = a->y - b->y;

  return z;
} /* }}} */

/* coords_plus(const VoronoiCoords*const, const VoronoiCoords*const) {{{ */
static inline VoronoiCoords
coords_plus(const VoronoiCoords *const a, const VoronoiCoords *const b)
{
  VoronoiCoords z;

  z.x = a->x + b->x;
  z.y = a->y + b->y;

  return z;
} /* }}} */

#endif /* VORONOI_H */


voronoi2d.c
/*
 * @(#) $Id: voronoi2d.c,v 1.15 2004/05/08 19:13:22 yeti Exp $
 * Generate 2D Voronoi diagrams in linear time.
 *
 * Copyright (C) 2001-2002,2004 Yeti (David Necas) <yeti@physics.muni.cz>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */
#define _GNU_SOURCE
#include <stdlib.h>
#include <math.h>
#include <glib.h>
#include "voronoi.h"

/* TODO: */
gdouble radial_factor;

static inline gdouble
angle(const VoronoiCoords *const r)
{
  return atan2(r->y, r->x);
}

static gint
vobj_angle_compare(gconstpointer x, gconstpointer y)
{
  gdouble xangle, yangle;

  xangle = ((VoronoiObject*)x)->angle;
  yangle = ((VoronoiObject*)y)->angle;

  if (xangle < yangle)
    return -1;
  if (xangle > yangle)
    return 1;
  return 0;
}

/* XXX: owner->ne requirements: NONE */
gdouble
value_flat(const VoronoiCoords *const point,
           const VoronoiObject *const owner)
{
  gdouble r;

  r = owner->random;

  return r;
}

/* XXX: owner->ne requirements: NONE */
gdouble
value_radial(const VoronoiCoords *const point,
             const VoronoiObject *const owner)
{
  VoronoiCoords dist;
  gdouble r;

  dist = coords_minus(point, &owner->pos);
  r = radial_factor*sqrt(DOTPROD_SS(dist, dist));

  return r;
}

/* XXX: owner->ne requirements: cyclic, neighbourized, segment angles */
gdouble
value_segmented(const VoronoiCoords *const point,
                const VoronoiObject *const owner)
{
  VoronoiCoords dist;
  gdouble r, phi;
  GSList *ne;

  ne = owner->ne;
  dist = coords_minus(point, &owner->pos);
  phi = angle(&dist);

  while ((phi >= VOBJ(ne)->angle)
         + (phi < VOBJ(ne->next)->angle)
         + (VOBJ(ne)->angle > VOBJ(ne->next)->angle) < 2)
    ne = ne->next;

  r = 2*DOTPROD_SS(dist, VOBJ(ne)->rel.v)/VOBJ(ne)->rel.d;

  return r;
}

/* XXX: owner->ne requirements: neighbourized */
gdouble
value_border(const VoronoiCoords *const point,
             const VoronoiObject *const owner)
{
  VoronoiCoords dist;
  gdouble r, r_min;
  GSList *ne;

  dist = coords_minus(point, &owner->pos);
  r_min = HUGE_VAL;

  for (ne = owner->ne; ne != NULL; ne = ne->next) {
    r = fabs(VOBJ(ne)->rel.d/2 - DOTPROD_SS(dist, VOBJ(ne)->rel.v))
        /sqrt(VOBJ(ne)->rel.d);
    if (r < r_min)
      r_min = r;
    if (ne->next == owner->ne)
      break;
  }

  r = 1 - 2*r_min*radial_factor;

  return r;
}

/* XXX: owner->ne requirements: NONE */
gdouble
value_second(const VoronoiCoords *const point,
             const VoronoiObject *const owner)
{
  VoronoiCoords dist;
  gdouble r, r_min;
  GSList *ne;

  r_min = HUGE_VAL;

  for (ne = owner->ne; ne != NULL; ne = ne->next) {
    dist = coords_minus(point, &VOBJ(ne)->pos);
    r = DOTPROD_SS(dist, dist);
    if (r < r_min)
      r_min = r;
    if (ne->next == owner->ne)
      break;
  }

  r = 1 - sqrt(r_min)*radial_factor;

  return r;
}

/* XXX: owner->ne requirements: NONE */
gdouble
value_sprod(const VoronoiCoords *const point,
            const VoronoiObject *const owner)
{
  VoronoiCoords dist, dist_min;
  gdouble r, r_min;
  GSList *ne;

  r_min = HUGE_VAL;
  dist_min.x = dist_min.y = 0.0;

  for (ne = owner->ne; ne != NULL; ne = ne->next) {
    dist = coords_minus(point, &VOBJ(ne)->pos);
    r = DOTPROD_SS(dist, dist);
    if (r < r_min) {
      r_min = r;
      dist_min = dist;
    }
    if (ne->next == owner->ne)
      break;
  }

  dist = coords_minus(&owner->pos, point);
  r = (DOTPROD_SS(dist, dist_min)*radial_factor*radial_factor + 0.5)/1.5;

  return r;
}

/* compute segment angles
   more precisely, VOBJ(ne)->angle will be set to start angle for segment
   from ne to ne->next (so end angle is in ne->next)
   XXX: ne0 requirements: cyclic and neighbourized */
void
compute_segment_angles(GSList *ne0)
{
  GSList *ne;
  VoronoiObject *p, *q;
  VoronoiCoords z;

  ne = ne0;
  do {
    p = VOBJ(ne);
    q = VOBJ(ne->next);
    z.x = p->rel.d * q->rel.v.y - q->rel.d * p->rel.v.y;
    z.y = q->rel.d * p->rel.v.x - p->rel.d * q->rel.v.x;
    q->angle = angle(&z);
    ne = g_slist_next(ne);
  } while (ne != ne0);
}

/* compute angles from rel.v relative coordinates
   XXX: ne0 requirements: neighbourized */
void
compute_straight_angles(GSList *ne0)
{
  GSList *ne;
  VoronoiObject *p;

  for (ne = ne0; ne; ne = g_slist_next(ne)) {
    p = VOBJ(ne);
    p->angle = angle(&p->rel.v);
    if (ne->next == ne0)
      return;
  }
}

/* compute relative positions and norms to center center
   XXX: ne0 requirements: NONE */
void
neighbourize(GSList *ne0, const VoronoiCoords *const center)
{
  GSList *ne;

  for (ne = ne0; ne; ne = g_slist_next(ne)) {
    VoronoiObject *p = VOBJ(ne);

    p->rel.v = coords_minus(&p->pos, center);
    p->rel.d = DOTPROD_SS(p->rel.v, p->rel.v);
    if (ne->next == ne0)
      return;
  }
}

/* return intersection time t for intersection of lines:

           r = linevec*t + start
     |r - a| = |r - b|
 */
static inline gdouble
intersection_time(const VoronoiCoords *const a, const VoronoiCoords *const b,
                  const VoronoiCoords *const linevec,
                  const VoronoiCoords *const start)
{
  VoronoiCoords p, q;
  gdouble s;

  /* line dividing a-neighbourhood and b-neighbourhood */
  q = coords_minus(b, a);
  p = coords_plus(b, a);

  /* FIXME: this is numerically unstable.  I didn't observed any problems even
     for sigma = 0.00001, nevertheless it is. (the probability something odd
     will happen fortunately decreases fast enough with increasing oddity of
     the thing)  almost zero denominator fortunately doesn't matter much, as
     long as numerator is not almost zero too.  then we are trying to find
     intersection point of two almost parallel lines, which is moreover quite
     near, so we want to know its position precisely.  I don't know how to do
     it with the least loss of precession.  sigh. */
  s = DOTPROD_SP(q, linevec);
  if (fabs(s) < 1e-14)
    s = 1e-14; /* better than nothing */
  return (DOTPROD_SS(q, p)/2 - DOTPROD_SP(q, start))/s;
}

/* being in point start owned by owner (XXX: this condition MUST be true)
   we want to get to point end and know our new owner
   returns the new owner; in addition, when next_safe is not NULL it stores
   there number of times we can repeat move along (end - start) vector still
   remaining in the new owner */
VoronoiObject*
move_along_line(const VoronoiObject *const owner,
                const VoronoiCoords *const start,
                const VoronoiCoords *const end, gint *next_safe)
{
  VoronoiCoords linevec;
  VoronoiObject *ow;
  GSList *ne, *nearest = NULL;
  gdouble t, t_min, t_back;

  ow = (VoronoiObject*)owner;
  linevec = coords_minus(end, start);
  t_back = 0;
  /* XXX: start must be owned by owner, or else strange things will happen */
  for ( ; ; ) {
    t_min = HUGE_VAL;
    ne = ow->ne;
    do {
      /* find intersection with border line between ow and ne
         FIXME: there apparently exist values t > t_back && t_back > t */
      t = intersection_time(&ow->pos, &VOBJ(ne)->pos, &linevec, start);
      if (t - t_back >= EPS && t < t_min) {
        t_min = t;
        nearest = ne;
      }
      ne = ne->next;
    } while (ne != ow->ne);

    /* no intersection inside the abscissa? then we are finished and can
       compute how many steps the same direction will remain in ow's
       neighbourhood */
    if (t_min > 1) {
      if (next_safe == NULL)
        return ow;
      if (t_min == HUGE_VAL)
        *next_safe = G_MAXINT;
      else
        *next_safe = floor(t_min) - 1;
      return ow;
    }

    /* otherwise nearest intersection determines a new owner */
    ow = VOBJ(nearest);
    t_back = t_min; /* time value showing we are going back */
  }
}

/* find and return owner of point point
   NB: this is crude and should not be used for anything else than initial
   grip, use move_along_line() then
   N: works for both cyclic and noncyclic ne-> */
VoronoiObject*
find_owner(GSList **squares,
           const gint wsq,
           const gint hsq,
           const VoronoiCoords *const point)
{
  GSList *ne;
  VoronoiObject *owner = NULL;
  VoronoiCoords dist;
  gint jx, jy;
  gint ix, iy;
  gdouble norm_min;

  jx = floor(point->x) + SQBORDER;
  jy = floor(point->y) + SQBORDER;

  /* scan the 25-neighbourhood
     XXX: I thought 9-neighbourhood was enough, but apparently it wasn't */
  norm_min = HUGE_VAL;
  for (ix = -2; ix <= 2; ix++) {
    if (jx + ix < 0 || jx + ix >= wsq + 2*SQBORDER)
      continue;
    for (iy = -2; iy <= 2; iy++) {
      if (jy + iy < 0 || jy + iy >= hsq + 2*SQBORDER)
        continue;
      for (ne = squares[(iy+jy)*(wsq + 2*SQBORDER) + ix+jx];
           ne != NULL;
           ne = ne->next) {
        dist = coords_minus(&VOBJ(ne)->pos, point);
        if (DOTPROD_SS(dist, dist) < norm_min) {
          norm_min = DOTPROD_SS(dist, dist);
          owner = VOBJ(ne);
        }
        if (ne->next == squares[(iy+jy)*(wsq + 2*SQBORDER) + ix+jx])
          break;
      }
    }
  }

  return owner;
}

/* return true iff point z (given as VoronoiLine) is shadowed by points a and b
   (XXX: all coordiantes are relative) */
static inline gboolean
in_shadow(const VoronoiLine *const a, const VoronoiLine *const b,
          const VoronoiCoords *const z)
{
  VoronoiCoords r, oa, ob, rz;
  gdouble s;

  /* artifical fix for periodic grids, because in Real World This Just Does Not
   * Happen
   * also mitigates the s == 0 case below, as the offending point would
   * be probably removed here */
  if (DOTPROD_SP(a->v, z) > 1.01*a->d
      && fabs(a->v.x * z->y - z->x * a->v.y) < 1e-12)
    return TRUE;
  if (DOTPROD_SP(b->v, z) > 1.01*b->d
      && fabs(b->v.x * z->y - z->x * b->v.y) < 1e-12)
    return TRUE;

  s = 2*(a->v.x * b->v.y - b->v.x * a->v.y);
  /* FIXME: what to do when s == 0 (or very near)??? */
  r.x = (a->d * b->v.y - b->d * a->v.y)/s;
  r.y = (b->d * a->v.x - a->d * b->v.x)/s;
  oa.x = -a->v.y;
  oa.y = a->v.x;
  ob.x = -b->v.y;
  ob.y = b->v.x;
  rz = coords_minus(z, &r);
  return (DOTPROD_SS(rz, rz) > DOTPROD_SS(r, r)
          && DOTPROD_PS(z, oa)*DOTPROD_SS(b->v, oa) > 0
          && DOTPROD_PS(z, ob)*DOTPROD_SS(a->v, ob) > 0);
}

static GSList*
extract_neighbourhood(GSList **squares,
                      gint wsq,
                      gint hsq,
                      VoronoiObject *p)
{
  GSList *ne = NULL;
  gint jx, jy;
  gint ix, iy;
  gint xwsq, xhsq;

  xwsq = wsq + 2*SQBORDER;
  xhsq = hsq + 2*SQBORDER;

  jx = floor(p->pos.x) + SQBORDER;
  jy = floor(p->pos.y) + SQBORDER;

  /* construct the 37-neighbourhood list */
  for (ix = -3; ix <= 3; ix++) {
    if (jx + ix < 0 || jx + ix >= xwsq)
      continue;
    for (iy = -3; iy <= 3; iy++) {
      if (abs(ix) == 3 && abs(iy) == 3)
        continue;
      if (jy + iy < 0 || jy + iy >= xhsq)
        continue;
      ne = g_slist_concat(g_slist_copy(squares[(iy+jy)*xwsq + ix+jx]), ne);
      if (ix == 0 && iy == 0)
        ne = g_slist_remove(ne, p);
    }
  }

  g_assert(ne != NULL);

  /* compute relative coordinates and angles */
  neighbourize(ne, &p->pos);
  compute_straight_angles(ne);

  return ne;
}

static GSList*
shadow_filter(GSList *ne)
{
  GSList *ne1, *ne2;
  gint notremoved;
  gint len;

  if (ne == NULL)
    return ne;

  /* make the list cyclic if it isn't already
   * (we have to unlink elements ourself then) */
  len = 1;
  for (ne2 = ne; ne2->next && ne2->next != ne; ne2 = g_slist_next(ne2))
    len++;
  if (len < 3)
    return ne;
  ne2->next = ne;

  /* remove objects shadowed by their ancestors and successors
     XXX: in non-degenerate case this is O(n*log(n)), but can be O(n*n) */
  ne1 = ne;
  notremoved = 0;
  do {
    ne2 = ne1->next;
    if (in_shadow(&VOBJ(ne1)->rel,
                  &VOBJ(ne2->next)->rel,
                  &VOBJ(ne2)->rel.v)) {
      ne1->next = ne2->next;
      g_slist_free_1(ne2);
      notremoved = 0;
      len--;
    }
    else {
      ne1 = ne2;
      notremoved++;
    }
  } while (notremoved < len && len > 2);

  return ne1; /* return cyclic list */
}

/* object-removal code, not used as the square side seems to be large enough
 * to make the number of objects artifically inserted into otherwise empty
 * squares really negligible => there's no need to simulate empty squares
 * by removing the objects later */
#if 0
/* duplicate possibly cyclic list as either cyclic or noncyclic */
static inline GSList*
duplicate_list(GSList *list,
               gboolean create_cyclic)
{
  GSList *l, *copy;

  if (!list)
    return NULL;

  copy = NULL;
  for (l = list; l->next && l->next != l; l = g_slist_next(l))
    copy = g_slist_prepend(copy, l->data);
  if (l != list)
    copy = g_slist_prepend(copy, l->data);

  l = copy;
  copy = g_slist_reverse(copy);
  if (create_cyclic)
    l->next = copy;

  return copy;
}

/* compute the number of squares with just one object
 * each square MUST contain some */
static gint
count_1squares(GSList **squares,
               const gint wsq,
               const gint hsq)
{
  gint i, count;
  gint xwsq, xhsq;

  xwsq = wsq + 2*SQBORDER;
  xhsq = hsq + 2*SQBORDER;

  count = 0;
  for (i = 0; i < xwsq*xhsq; i++) {
    if (!squares[i]->next)
      count++;
  }

  return count;
}

static void
remove_object(VoronoiObject *obj)
{
  GSList *necopy, *ne, *nene, *ne2;

  g_return_if_fail(obj->ne);
  /* neighbours of the removed object */
  necopy = duplicate_list(obj->ne, FALSE);
  for (ne = necopy; ne; ne = g_slist_next(ne)) {
    VoronoiObject *neighbour = VOBJ(ne);

    /* join both neighbourhoods */
    nene = g_slist_concat(duplicate_list(neighbour->ne, FALSE),
                          g_slist_copy(necopy));
    /* remove object and neighbour itself */
    nene = g_slist_remove_all(nene, obj);
    nene = g_slist_remove_all(nene, neighbour);
    /* neighbourize */
    neighbourize(nene, &neighbour->pos);
    compute_straight_angles(nene);
    nene = g_slist_sort(nene, &vobj_angle_compare);
    /* remove duplicities */
    for (ne2 = nene; ne2 && ne2->next; ne2 = g_slist_next(ne2)) {
      if (ne2->data == ne2->next->data) {
        GSList *nex;

        nex = ne2->next;
        ne2->next = nex->next;
        g_slist_free_1(nex);
      }
    }
    /* shadow filter and replace the neighbourhood */
    nene = shadow_filter(nene);
    g_slist_free(neighbour->ne);
    neighbour->ne = nene;
  }
  g_slist_free(necopy);
}

void
remove_objects(GSList **squares,
               const gint wsq,
               const gint hsq,
               gboolean htil,
               gboolean vtil)
{
  gdouble n1;
  gint nremove;
  gint xwsq, xhsq;

  xwsq = wsq + 2*SQBORDER;
  xhsq = hsq + 2*SQBORDER;

  n1 = count_1squares(squares, wsq, hsq);
  nremove = n1*n1/(xwsq*xhsq);
  fprintf(stderr, "n1 = %.0f, N = %d, nremove = %d\n",
          n1, xwsq*xhsq, nremove);
}
#endif

void
find_voronoi_neighbours_iter(GSList **squares,
                             const gint wsq,
                             const gint hsq,
                             const gint iter)
{
  GSList *this;

  for (this = squares[iter]; this; this = g_slist_next(this)) {
    VoronoiObject *obj = VOBJ(this);

    obj->ne = extract_neighbourhood(squares, wsq, hsq, obj);
    obj->ne = g_slist_sort(obj->ne, &vobj_angle_compare);
    obj->ne = shadow_filter(obj->ne);
  }
}

void
find_voronoi_neighbours(GSList **squares,
                        const gint wsq,
                        const gint hsq)
{
  gint xwsq, xhsq, i;

  xwsq = wsq + 2*SQBORDER;
  xhsq = hsq + 2*SQBORDER;

  for (i = 0; i < xwsq*xhsq; i++)
    find_voronoi_neighbours_iter(squares, wsq, hsq, i);
}

void
destroy_neighbourhoods(GSList **squares,
                       const gint wsq,
                       const gint hsq)
{
  gint xwsq, xhsq, i;
  GSList *this, *ne;

  xwsq = wsq + 2*SQBORDER;
  xhsq = hsq + 2*SQBORDER;

  for (i = 0; i < xwsq*xhsq; i++) {
    for (this = squares[i]; this; this = g_slist_next(this)) {
      ne = VOBJ(this)->ne->next;
      VOBJ(this)->ne->next = NULL; /* break cycle */
      g_slist_free(ne);
    }
  }
}

